File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Collide/CrateCollide/HealCrateCollide.cpp
// FILE: MoneyCrateCollide.cpp ///////////////////////////////////////////////////////////////////////
// Author: Graham Smallwood, March 2002
// Desc:   A crate that heals everything owned by the collider
///////////////////////////////////////////////////////////////////////////////////////////////////

// INCLUDES ///////////////////////////////////////////////////////////////////////////////////////
#include "PreRTS.h"	// This must go first in EVERY cpp file int the GameEngine
#include "Common/AudioEventRTS.h"
#include "Common/MiscAudio.h"
#include "Common/Player.h"
#include "Common/Xfer.h"
#include "GameLogic/Object.h"
#include "GameLogic/Module/HealCrateCollide.h"

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
HealCrateCollide::HealCrateCollide( Thing *thing, const ModuleData* moduleData ) : CrateCollide( thing, moduleData )
{

} 

//-------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------
HealCrateCollide::~HealCrateCollide( void )
{

}  

//-------------------------------------------------------------------------------------------------
Bool HealCrateCollide::executeCrateBehavior( Object *other )
{
	Player* cratePlayer = other->getControllingPlayer();
	cratePlayer->healAllObjects();

	//Play a crate pickup sound.
	AudioEventRTS soundToPlay = TheAudio->getMiscAudio()->m_crateHeal;
	soundToPlay.setPosition( other->getPosition() );
	TheAudio->addAudioEvent(&soundToPlay);
	
	return TRUE;
}

// ------------------------------------------------------------------------------------------------
/** CRC */
// ------------------------------------------------------------------------------------------------
void HealCrateCollide::crc( Xfer *xfer )
{

	// extend base class
	CrateCollide::crc( xfer );

}  // end crc

// ------------------------------------------------------------------------------------------------
/** Xfer method
	* Version Info:
	* 1: Initial version */
// ------------------------------------------------------------------------------------------------
void HealCrateCollide::xfer( Xfer *xfer )
{

	// version
	XferVersion currentVersion = 1;
	XferVersion version = currentVersion;
	xfer->xferVersion( &version, currentVersion );

	// extend base class
	CrateCollide::xfer( xfer );

}  // end xfer

// ------------------------------------------------------------------------------------------------
/** Load post process */
// ------------------------------------------------------------------------------------------------
void HealCrateCollide::loadPostProcess( void )
{

	// extend base class
	CrateCollide::loadPostProcess();

}  // end loadPostProcess

File: GeneralsRust/Code/GameEngine/GameLogic/src/object/collide/crate_collide/heal_crate_collide.rs
//! Heal Crate Collision Module
//!
//! A crate that heals every object owned by the player who collects it.

use super::super::{CollisionError, Coord3D, GameObject};
use super::crate_collide::{CrateCollide, CrateCollideBehavior, CrateCollideModuleData};
use crate::common::*;
use crate::helpers::TheAudio;
use crate::object::collide::crate_collide::*;

/// Configuration data for HealCrateCollide.
///
/// C++ exposes only inherited CrateCollide module data for this module.
#[derive(Debug, Clone)]
pub struct HealCrateCollideModuleData {
    /// Base crate collision data
    pub base: CrateCollideModuleData,
}

impl HealCrateCollideModuleData {
    pub fn new() -> Self {
        Self {
            base: CrateCollideModuleData::new(),
        }
    }
}

impl Default for HealCrateCollideModuleData {
    fn default() -> Self {
        Self::new()
    }
}

/// Heal Crate Collide implementation.
pub struct HealCrateCollide {
    /// Base crate collision functionality
    base_crate: CrateCollide,
    /// Module-specific configuration
    module_data: HealCrateCollideModuleData,
}

impl HealCrateCollide {
    pub fn new(object_id: ObjectId, module_data: HealCrateCollideModuleData) -> Self {
        Self {
            base_crate: CrateCollide::new(object_id, module_data.base.clone()),
            module_data,
        }
    }

    pub fn get_module_data(&self) -> &HealCrateCollideModuleData {
        &self.module_data
    }

    /// Execute the C++ heal-crate behavior.
    pub fn execute_healing(&mut self, other: &dyn GameObject) -> Result<bool, CollisionError> {
        let Some(other_handle) = other.as_object_handle() else {
            return Ok(false);
        };
        let Some(player) = other_handle
            .read()
            .map_err(|_| CollisionError::InvalidObject("Failed to lock collector".to_string()))?
            .get_controlling_player()
        else {
            return Ok(false);
        };

        if let Ok(mut player_guard) = player.write() {
            player_guard.heal_all_objects();
        }

        self.play_heal_audio(&other.get_position());
        Ok(true)
    }

    fn play_heal_audio(&self, position: &Coord3D) {
        if let Some(audio) = TheAudio::get() {
            let event = TheAudio::get_misc_audio().crate_heal.clone();
            let mut audio_event = crate::common::audio::AudioEventRts::new(event.sound_type);
            audio_event.set_position(&(position.x, position.y, position.z));
            audio.add_audio_event(&audio_event);
        }
    }
}

impl CrateCollideBehavior for HealCrateCollide {
    fn execute_crate_behavior(&mut self, other: &dyn GameObject) -> Result<bool, CollisionError> {
        self.execute_healing(other)
    }

    fn is_valid_to_execute(&self, other: &dyn GameObject) -> bool {
        self.base_crate.is_valid_to_execute(other)
    }
}

/// Factory for creating HealCrateCollide modules.
pub struct HealCrateCollideFactory;

impl HealCrateCollideFactory {
    pub fn create(object_id: ObjectId) -> HealCrateCollide {
        let data = HealCrateCollideModuleData::new();
        HealCrateCollide::new(object_id, data)
    }

    pub fn create_with_config(
        object_id: ObjectId,
        config: HealCrateCollideModuleData,
    ) -> HealCrateCollide {
        HealCrateCollide::new(object_id, config)
    }
}

impl game_engine::common::system::Snapshotable for HealCrateCollide {
    fn crc(&self, xfer: &mut dyn game_engine::common::system::Xfer) -> Result<(), String> {
        self.base_crate.crc(xfer)
    }

    fn xfer(&mut self, xfer: &mut dyn game_engine::common::system::Xfer) -> Result<(), String> {
        let mut version: u8 = 1;
        xfer.xfer_version(&mut version, 1)
            .map_err(|err| err.to_string())?;
        self.base_crate.xfer(xfer)
    }

    fn load_post_process(&mut self) -> Result<(), String> {
        self.base_crate.load_post_process()
    }
}
